### PR TITLE
Fix some flaky tests in test_torch.py and test_nn.py

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Fixed test:
- `test_is_nonzero`, this is asserting exact match, which is flaky when `TORCH_SHOW_CPP_STACKTRACES=1`, I changed this to non-exact assert
- `test_pinverse` TF32
- `test_symeig` TF32
- `test_triangular_solve_batched_many_batches_cpu_float64` precision on CPU BLAS
- `test_qr` TF32, as well as the tensor factory forgets a `dtype=dtype`
- `test_lu` TF32
- `ConvTranspose2d` TF32
- `Conv3d_1x1x1_no_bias` TF32
- `Transformer*` TF32

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -9699,9 +9694,13 @@ def test_rpow(self, device):
     @skipCPUIfNoLapack
     @dtypesIfCPU(torch.float32, torch.float64, torch.complex64, torch.complex128)
     @dtypesIfCUDA(torch.float32, torch.float64)
+    @tf32_on_and_off()
     def test_symeig(self, device, dtype):
         from torch.testing._internal.common_utils import random_hermitian_matrix
 
+        is_tf32 = (dtype == torch.float32 and tf32_is_not_fp32() and torch.backends.cuda.matmul.allow_tf32)
+        recon_tol = 5e-3 if is_tf32 else 1e-8
```
Comment (mruberry) [inline test/test_torch.py:9702]: What does "recon_tol" mean?

**Comment by mruberry** [comment]
Test failures seem real.

File: test/test_torch.py (hunk 1 above)
Comment (zasdfgbnm) [reply test/test_torch.py]: I renamed it to `reconstruct_tol`, which is the tolerance for reconstructing from eigen decomposition

**Comment by zasdfgbnm** [comment]
@mruberry Yes, test failures are real. The original `test_qr` has multiple bugs, and I need to fix all of them to make it pass.

**Comment by zasdfgbnm** [comment]
@mruberry @ngimel This is ready

**Comment by zasdfgbnm** [comment]
How can this PR get merge conflicts so often?

**Comment by zasdfgbnm** [comment]
ping @ngimel  @mruberry

File: test/test_torch.py [hunk 2]
```diff
@@ -7025,7 +7020,7 @@ def test_inverse_many_batches_helper(b, n):
         test_inverse_many_batches_helper(3, 512)
         test_inverse_many_batches_helper(64, 64)
 
-    @precisionOverride({torch.float32: 1e-3, torch.complex64: 1e-3})
+    @precisionOverride({torch.float32: 5e-3, torch.complex64: 1e-3})
```
Comment (ngimel) [inline test/test_torch.py:7023]: why do you need new tolerance?

File: test/test_torch.py [hunk 3]
```diff
@@ -7036,11 +7031,13 @@ def test_pinverse(self, device, dtype):
         def run_test(M):
             # Testing against definition for pseudo-inverses
             MPI = torch.pinverse(M)
+            MPI_ = MPI.cpu().numpy()
+            M_ = M.cpu().numpy()
```
Comment (ngimel) [inline test/test_torch.py:7035]: would just M.cpu() and torch.matmul work to avoid tf32, or do you need np.matmul?

File: test/test_torch.py [hunk 4]
```diff
@@ -15694,6 +15703,7 @@ def run_subtest(matrix_size, batches, device, pivot, singular=False, a=None):
                     rows, columns = matrix_size
                 if a is None:
                     a = random_matrix(rows, columns, *batches, **dict(singular=singular, dtype=dtype)).to(device)
+                a_ = a.cpu().numpy()
```
Comment (ngimel) [inline test/test_torch.py:15706]: I don't see where a_ is used - maybe github is not rendering.

File: test/test_torch.py (hunk 3 above)
Comment (zasdfgbnm) [reply test/test_torch.py:7035]: Did't we decide to use numpy to do the computation everywhere?

File: test/test_torch.py (hunk 2 above)
Comment (zasdfgbnm) [reply test/test_torch.py:7023]: I don't know, it is just failing, even if I added `NoTF32Guard` to this op.

File: test/test_torch.py (hunk 4 above)
Comment (zasdfgbnm) [reply test/test_torch.py:15706]: You are right, it is not used

File: test/test_torch.py (hunk 2 above)
Comment (zasdfgbnm) [reply test/test_torch.py:7023]: It could be different rounding errors of CUDA BLAS vs CPU BLAS

File: test/test_torch.py (hunk 4 above)
Comment (zasdfgbnm) [reply test/test_torch.py]: fixed

File: test/test_torch.py (hunk 3 above)
Comment (mruberry) [reply test/test_torch.py:7035]: I don't know if we're that prescriptive (yet), but this should be fine. We definitely want to use NumPy for reference impls.

**Review by mruberry — APPROVED** [review]
Thanks @zasdfgbnm!

**Comment by mruberry** [comment]
Sorry @zasdfgbnm, this needs a rebase to resolve the test_torch.py merge conflict.

**Pull request closed without merging**